### PR TITLE
Enhance benchmarking

### LINKED ISSUE #12331: Divan benchmarking tracking
### Question

Currently, Divan does not any exposed file output interface. There is a [PR](https://github.com/nvzqz/divan/pull/42) but it has yet to be merged ☹️. This looks quite straightforward to use, so I took some time to create [nushell script](https://gist.github.com/FilipAndersson245/e89b3870679e1bf023d8c828fedd0a56) that processes this and returns back a simple table that can be used for inspection. this script takes 2 json an old and a new which is the output from the divan fork.
If we want to use this now we would have to specify the GitHub fork until it is merged.

also as @devyn said there is still some noise in the data, so I suggest we add a minimum time of at least 1 sec for the benches, as this seem the most prominent to be noisy in extremely fast benches.

# Markdown preview
|Name|Old|New|Relative|Note|
|-|-|-|-|-|
|load_standard_lib|7ms 535µs 7ns|7ms 431µs 498ns|1.01| |
|decoding_benchmarks / json_decode / (100, 5)|266µs 386ns|262µs 787ns|1.01| |
|decoding_benchmarks / json_decode / (10000, 15)|118ms 477µs 833ns|118ms 146µs 34ns|1| |
|decoding_benchmarks / msgpack_decode / (100, 5)|167µs 965ns|169µs 179ns|0.99| |
|decoding_benchmarks / msgpack_decode / (10000, 15)|63ms 74µs 606ns|61ms 139µs 748ns|1.03| |
|encoding_benchmarks / json_encode / (100, 5)|33µs 785ns|33µs 880ns|1| |
|encoding_benchmarks / json_encode / (10000, 15)|9ms 239µs 443ns|8ms 612µs 413ns|1.07| |
|encoding_benchmarks / msgpack_encode / (100, 5)|22µs 570ns|16µs 731ns|1.35|🟢|
|encoding_benchmarks / msgpack_encode / (10000, 15)|4ms 326µs 401ns|4ms 241µs 439ns|1.02| |
|eval_benchmarks / eval_default_config|1ms 917µs 747ns|1ms 847µs 826ns|1.04| |
|eval_benchmarks / eval_default_env|320µs 923ns|310µs 102ns|1.03| |
|eval_commands / each / 1|104µs 802ns|101µs 138ns|1.04| |
|eval_commands / each / 5|326µs 558ns|320µs 778ns|1.02| |
|eval_commands / each / 10|603µs 770ns|592µs 775ns|1.02| |
|eval_commands / each / 100|5ms 481µs 600ns|5ms 459µs 470ns|1| |
|eval_commands / each / 1000|54ms 624µs 31ns|54ms 494µs 478ns|1| |
|eval_commands / for_range / 1|105µs 87ns|101µs 737ns|1.03| |
|eval_commands / for_range / 5|322µs 243ns|318µs 18ns|1.01| |
|eval_commands / for_range / 10|588µs 464ns|588µs 429ns|1| |
|eval_commands / for_range / 100|5ms 450µs 9ns|5ms 472µs 521ns|1| |
|eval_commands / for_range / 1000|54ms 163µs 760ns|53ms 905µs 429ns|1| |
|eval_commands / interleave / 100|833µs 990ns|487µs 732ns|1.71|🟢🟢|
|eval_commands / interleave / 1000|8ms 84µs 884ns|4ms 241µs 476ns|1.91|🟢🟢|
|eval_commands / interleave / 10000|75ms 457µs 177ns|40ms 962µs 963ns|1.84|🟢🟢|
|eval_commands / interleave_with_ctrlc / 100|484µs 412ns|487µs 282ns|0.99| |
|eval_commands / interleave_with_ctrlc / 1000|4ms 360µs 723ns|4ms 296µs 922ns|1.01| |
|eval_commands / interleave_with_ctrlc / 10000|76ms 192µs 511ns|42ms 758µs 658ns|1.78|🟢🟢|
|eval_commands / par_each_1t / 1|137µs 433ns|118µs 323ns|1.16| |
|eval_commands / par_each_1t / 5|351µs 823ns|335µs 828ns|1.05| |
|eval_commands / par_each_1t / 10|639µs 291ns|609µs 650ns|1.05| |
|eval_commands / par_each_1t / 100|5ms 459µs 59ns|5ms 440µs 965ns|1| |
|eval_commands / par_each_1t / 1000|53ms 696µs 994ns|53ms 527µs 860ns|1| |
|eval_commands / par_each_2t / 1|148µs 38ns|125µs 193ns|1.18| |
|eval_commands / par_each_2t / 5|252µs 837ns|234µs 40ns|1.08| |
|eval_commands / par_each_2t / 10|373µs 19ns|348µs 609ns|1.07| |
|eval_commands / par_each_2t / 100|2ms 840µs 175ns|2ms 853µs 36ns|1| |
|eval_commands / par_each_2t / 1000|27ms 104µs 871ns|26ms 978µs 33ns|1| |
|parser_benchmarks / parse_default_config_file|1ms 394µs 410ns|1ms 150µs 163ns|1.21| |
|parser_benchmarks / parse_default_env_file|235µs 511ns|217µs 441ns|1.08| |
|record / create / 1|20µs 300ns|19µs 800ns|1.03| |
|record / create / 10|29µs 716ns|28µs 321ns|1.05| |
|record / create / 100|129µs 288ns|123µs 733ns|1.04| |
|record / create / 1000|1ms 59µs 126ns|1ms 60µs 171ns|1| |
|record / flat_access / 1|12µs 685ns|12µs 625ns|1| |
|record / flat_access / 10|12µs 130ns|11µs 901ns|1.02| |
|record / flat_access / 100|18µs 121ns|18µs 840ns|0.96| |
|record / flat_access / 1000|97µs 373ns|103µs 642ns|0.94| |
|record / nest_access / 1|11µs 10ns|10µs 590ns|1.04| |
|record / nest_access / 2|12µs 445ns|13µs 740ns|0.91| |
|record / nest_access / 4|12µs 935ns|15µs 60ns|0.86| |
|record / nest_access / 8|19µs 35ns|16µs 495ns|1.15| |
|record / nest_access / 16|24µs 320ns|27µs 771ns|0.88| |
|record / nest_access / 32|51µs 351ns|50µs 831ns|1.01| |
|record / nest_access / 64|155µs 689ns|153µs 468ns|1.01| |
|record / nest_access / 128|579µs 84ns|553µs 609ns|1.05| |
|table / create / 1|23µs 405ns|23µs 35ns|1.02| |
|table / create / 10|35µs 740ns|35µs 565ns|1| |
|table / create / 100|141µs 573ns|140µs 953ns|1| |
|table / create / 1000|1ms 206µs 674ns|1ms 218µs 645ns|0.99| |
|table / get / 1|15µs 155ns|15µs 120ns|1| |
|table / get / 10|21µs 490ns|18µs 245ns|1.18| |
|table / get / 100|45µs 946ns|45µs 256ns|1.02| |
|table / get / 1000|305µs 117ns|307µs 532ns|0.99| |
|table / select / 1|14µs 670ns|14µs 150ns|1.04| |
|table / select / 10|20µs 830ns|20µs 110ns|1.04| |
|table / select / 100|100µs 753ns|72µs 782ns|1.38|🟢|
|table / select / 1000|601µs 265ns|603µs 314ns|1| |

### Additional context and details

Using the PR as a base you can save the output with this
`cargo bench -- --file="./mybench.json"`

and the script is run by `(compare-benchmarks "./old.json" "./new.json") | to md | save result.md`

### PR DESCRIPTION
**Pull request by FilipAndersson245**
# Description

closes #12331

This PR does 3 things
- changes divan to a specific commit to enable file exports.
- sets a min_time for each current bench, this to reduce noise for fast benches.
- Adds a few new commands to toolkit.nu,
  - format-benchmark-data: convert the benchmark data from divan to a simple flat table.
  - compare-benchmarks: takes 2 different benchmarks and try to generate a simple report comparing them.

image

# User-Facing Changes

# Tests + Formatting

# After Submitting

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -211,7 +211,7 @@ nix = { workspace = true, default-features = false, features = [
 nu-test-support = { path = "./crates/nu-test-support", version = "0.91.1" }
 assert_cmd = "2.0"
 dirs-next = { workspace = true }
-divan = "0.1.14"
+divan = { git = "https://github.com/OliverKillane/divan.git", rev = "c54ac74" }
```
Comment (fdncred) [inline Cargo.toml:214]: I thought that git dependencies prevent crates from being published. Not sure if that's the same for [dev-dependencies] though.

**Comment by fdncred** [comment]
Just one comment above but I'm fine with the rest of the PR. I just don't want this change to cause our publishing to puke.

File: Cargo.toml (hunk 1 above)
Comment (FilipAndersson245) [reply Cargo.toml:214]: I am not sure how exactly we publish releases, how would I test this out.

I belive it should not affect, as it should not be built when run with cargo run --release

**Comment by FilipAndersson245** [comment]
If someone in core could help me test this out it would be appreciated 👏

**Comment by FilipAndersson245** [comment]
I changed the default comparason metric from `median` to `fastest` based on [this](https://nikolaivazquez.com/blog/divan/#robust-benchmarking-paper)

**Comment by FilipAndersson245** [comment]
Im trying to rewrite this PR and instead using the main branch of divan and instead try to parse down the graphical output from divan.

**Comment by kubouch** [comment]
I checked it, and it seems we can't publish a crate with unpublished dependencies, so the git link won't work.

It is a bit unfortunate that divan development has slowed down. Last commit was 2 months ago and divan's PR implementing file output hasn't yet received any maintainer comments despite being open since January. We can roll with your parsing script, but if divan ends up being unmaintained, we might need to switch back to criterion or something that's at least somewhat maintained.

**Comment by FilipAndersson245** [comment]
> I checked it, and it seems we can't publish a crate with unpublished dependencies, so the git link won't work.
> 
> It is a bit unfortunate that divan development has slowed down. Last commit was 2 months ago and divan's PR implementing file output hasn't yet received any maintainer comments despite being open since January. We can roll with your parsing script, but if divan ends up being unmaintained, we might need to switch back to criterion or something that's at least somewhat maintained.

I have a new script that parses the raw output, in the oven, probably try to get it in this PR, so we don`t have to use a PR

**Comment by FilipAndersson245** [comment]
@fdncred @kubouch I have tried out a new approach that relies on parsing the visual printout of divan, I think its decent, but may break if they change their format.

# The new markdown display format
|Name|Old|New|Relative|Note|
|-|-|-|-|-|
| load_standard_lib|8ms 502µs|8ms 502µs|1| |
| decoding_benchmarks|||||
|&ensp; json_decode|||||
|&ensp;&ensp; (100, 5)|406µs 200ns|406µs 200ns|1| |
|&ensp;&ensp; (10000, 15)|157ms 100µs|157ms 100µs|1| |
|&ensp; msgpack_decode|||||
|&ensp;&ensp; (100, 5)|179µs 100ns|179µs 100ns|1| |
|&ensp;&ensp; (10000, 15)|63ms 720µs|63ms 720µs|1| |
| encoding_benchmarks|||||
|&ensp; json_encode|||||
|&ensp;&ensp; (100, 5)|87µs 280ns|87µs 280ns|1| |
|&ensp;&ensp; (10000, 15)|21ms 50µs|21ms 50µs|1| |
|&ensp; msgpack_encode|||||
|&ensp;&ensp; (100, 5)|35µs 400ns|35µs 400ns|1| |
|&ensp;&ensp; (10000, 15)|9ms 212µs|9ms 212µs|1| |
| eval_benchmarks|||||
|&ensp; eval_default_config|2ms 216µs|2ms 216µs|1| |
|&ensp; eval_default_env|347µs 400ns|347µs 400ns|1| |
| eval_commands|||||
|&ensp; each|||||
|&ensp;&ensp; 1|51µs|51µs|1| |
|&ensp;&ensp; 5|267µs 400ns|267µs 400ns|1| |
|&ensp;&ensp; 10|489µs 400ns|489µs 400ns|1| |
|&ensp;&ensp; 100|5ms 339µs|5ms 339µs|1| |
|&ensp;&ensp; 1000|53ms 560µs|53ms 560µs|1| |
|&ensp; for_range|||||
|&ensp;&ensp; 1|51µs 320ns|51µs 320ns|1| |
|&ensp;&ensp; 5|266µs|266µs|1| |
|&ensp;&ensp; 10|532µs 100ns|532µs 100ns|1| |
|&ensp;&ensp; 100|5ms 247µs|5ms 247µs|1| |
|&ensp;&ensp; 1000|53ms 150µs|53ms 150µs|1| |
|&ensp; interleave|||||
|&ensp;&ensp; 100|420µs 400ns|420µs 400ns|1| |
|&ensp;&ensp; 1000|3ms 943µs|3ms 943µs|1| |
|&ensp;&ensp; 10000|38ms 730µs|38ms 730µs|1| |
|&ensp; interleave_with_ctrlc|||||
|&ensp;&ensp; 100|421µs 100ns|421µs 100ns|1| |
|&ensp;&ensp; 1000|3ms 784µs|3ms 784µs|1| |
|&ensp;&ensp; 10000|40ms 650µs|40ms 650µs|1| |
|&ensp; par_each_1t|||||
|&ensp;&ensp; 1|69µs 340ns|69µs 340ns|1| |
|&ensp;&ensp; 5|285µs 400ns|285µs 400ns|1| |
|&ensp;&ensp; 10|517µs 900ns|517µs 900ns|1| |
|&ensp;&ensp; 100|5ms 405µs|5ms 405µs|1| |
|&ensp;&ensp; 1000|53ms 300µs|53ms 300µs|1| |
|&ensp; par_each_2t|||||
|&ensp;&ensp; 1|77µs 750ns|77µs 750ns|1| |
|&ensp;&ensp; 5|189µs 200ns|189µs 200ns|1| |
|&ensp;&ensp; 10|304µs 700ns|304µs 700ns|1| |
|&ensp;&ensp; 100|2ms 785µs|2ms 785µs|1| |
|&ensp;&ensp; 1000|26ms 750µs|26ms 750µs|1| |
| parser_benchmarks|||||
|&ensp; parse_default_config_file|1ms 397µs|1ms 397µs|1| |
|&ensp; parse_default_env_file|247µs 100ns|247µs 100ns|1| |
| record|||||
|&ensp; create|||||
|&ensp;&ensp; 1|20µs 560ns|20µs 560ns|1| |
|&ensp;&ensp; 10|31µs 240ns|31µs 240ns|1| |
|&ensp;&ensp; 100|138µs 300ns|138µs 300ns|1| |
|&ensp;&ensp; 1000|1ms 229µs|1ms 229µs|1| |
|&ensp; flat_access|||||
|&ensp;&ensp; 1|11µs 210ns|11µs 210ns|1| |
|&ensp;&ensp; 10|11µs 890ns|11µs 890ns|1| |
|&ensp;&ensp; 100|19µs 100ns|19µs 100ns|1| |
|&ensp;&ensp; 1000|99µs 590ns|99µs 590ns|1| |
|&ensp; nest_access|||||
|&ensp;&ensp; 1|11µs 640ns|11µs 640ns|1| |
|&ensp;&ensp; 2|11µs 950ns|11µs 950ns|1| |
|&ensp;&ensp; 4|13µs 820ns|13µs 820ns|1| |
|&ensp;&ensp; 8|18µs 470ns|18µs 470ns|1| |
|&ensp;&ensp; 16|26µs 980ns|26µs 980ns|1| |
|&ensp;&ensp; 32|57µs 960ns|57µs 960ns|1| |
|&ensp;&ensp; 64|178µs 600ns|178µs 600ns|1| |
|&ensp;&ensp; 128|678µs 300ns|678µs 300ns|1| |
| table|||||
|&ensp; create|||||
|&ensp;&ensp; 1|23µs 710ns|23µs 710ns|1| |
|&ensp;&ensp; 10|35µs 870ns|35µs 870ns|1| |
|&ensp;&ensp; 100|156µs 900ns|156µs 900ns|1| |
|&ensp;&ensp; 1000|1ms 368µs|1ms 368µs|1| |
|&ensp; get|||||
|&ensp;&ensp; 1|16µs 340ns|16µs 340ns|1| |
|&ensp;&ensp; 10|19µs 580ns|19µs 580ns|1| |
|&ensp;&ensp; 100|50µs 90ns|50µs 90ns|1| |
|&ensp;&ensp; 1000|335µs 500ns|335µs 500ns|1| |
|&ensp; select|||||
|&ensp;&ensp; 1|15µs 480ns|15µs 480ns|1| |
|&ensp;&ensp; 10|22µs 30ns|22µs 30ns|1| |
|&ensp;&ensp; 100|79µs 50ns|79µs 50ns|1| |
|&ensp;&ensp; 1000|646µs 300ns|646µs 300ns|1| |
||||||
 
# Nushell record look
image

**Comment by kubouch** [comment]
Would it be possible to modify the output to not contain the empty entries? It could be made into a tree structure, or added as a prefix, e.g., `encoding_benchmarks::json_decode::(100, 5)`, `encoding_benchmarks::json_decode::(10000, 5)`, etc.

Otherwise, I think we could roll with this and see how it goes with divan's development in the future.

**Comment by FilipAndersson245** [comment]
I think this works fine, i am not to sure how to format the data correctly for that.

**Comment by kubouch** [comment]
It looks fine, but it's harder to do any processing with it. But we can keep it as-is, since it's still an improvement over having no parsing. The suggestion I made can be added later as another postprocessing step, for example.

**Comment by FilipAndersson245** [comment]
I am thinking on what you said @kubouch and maybe Criterion would be better, in the least part cause it so standardized in the community. And would most likely live on in some form in the longterm.

**Comment by kubouch** [comment]
Yeah, divan seems really cool, but is more uncertain than criterion. We could switch back to criterion and see what happens with divan.

**Comment by FilipAndersson245** [comment]
> Yeah, divan seems really cool, but is more uncertain than criterion. We could switch back to criterion and see what happens with divan.

My only problem is that the maintenance, status of Criterion is quite uncertain also, as the main contributor have not committed in almost 2 years. I am becoming inspired to write my own 😎.

**Comment by kubouch** [comment]
Haha, that's the way :-D. Maybe it would be worth checking with the divan maintainer and perhaps offering help with the maintenance instead of starting from scratch.

The main plus of criterion is being able to get some structured data out of it (e.g., JSON). I've been playing with ideas of how to set up continuous benchmark tracking, but it's quite essential to be able to reliably output structured data. If we could get the JSON PR landed to divan, we could just continue using divan as long as it works in the current form.

**Comment by FilipAndersson245** [comment]
I have create an alternative PR where i migrate the benchmarks back to criterion, but after the talk jon had, that was linked in the weekly meeting I have started to try out tango.rs and it seem very interesting, it works by compiling two versions of a program and comparing them. I may explore if that would be a better alternative. As I am not that sold on criterion  either as it's slow and not as robust as it gives the shine of.

**Comment by fdncred** [comment]
where are we at on the benchmarks? does this pr need to land for the alternate or are we chosing between the two?

**Comment by FilipAndersson245** [comment]
It depend, I have done some talks with the author of Tango, as I had some initial difficulties using it in nutshell but it seems to be my fault for failing to detect our existing build.rs

We have 3 options,
1. Stay with what we have now. I think this works decent, only sad part is that we are a bit locked out of making nice data collecting.
2. I have created an alternative PR, that moves all existing benchmarks from this back to criterion.
3. We move to tango, which still is quite actively developed, but I fell this pair-benchmarking provide something quite nice in terms of evaluating performance.

I am personally quite annoyed about how slow criterion is and how it says its statistically significant changes but I fell quite often I see random multi digit changes between run. Tango seem both faster and more stable as what I can see, It also seem to already support file exports, And is actively maintained.

I was initially made aware of Tango from Jon Gjengset latest Rust Talk.

**Pull request closed without merging**